src/box/txn.c

```diff
@@ -573,38 +573,42 @@ txn_begin_in_engine(struct engine *engine, struct txn *txn)
 {
 	if (engine->flags & ENGINE_BYPASS_TX)
 		return 0;
+	assert(engine->id < MAX_TX_ENGINE_COUNT);
 	if (txn->engines[engine->id] != NULL) {
 		assert(txn_has_flag(txn, TXN_IS_STARTED_IN_ENGINE));
 		assert(txn->engines[engine->id] == engine);
 		return 0;
 	}
-	if (txn_has_flag(txn, TXN_IS_STARTED_IN_ENGINE)) {
-		if ((engine->flags & ENGINE_SUPPORTS_CROSS_ENGINE_TX) == 0) {
-			diag_set(ClientError, ER_CROSS_ENGINE_TRANSACTION,
-				 engine->name);
-			return -1;
-		}
-		if (!txn_has_flag(txn, TXN_SUPPORTS_MULTI_ENGINE)) {
-			const char *name = NULL;
-			for (size_t i = 0; i < MAX_TX_ENGINE_COUNT; i++) {
-				if (txn->engines[i] != NULL &&
-				    (txn->engines[i]->flags &
-				     ENGINE_SUPPORTS_CROSS_ENGINE_TX) == 0) {
-					name = txn->engines[i]->name;
-					break;
-				}
+	bool has_mvcc = (engine->flags & ENGINE_SUPPORTS_MVCC) != 0;
+	if (unlikely(txn_has_flag(txn, TXN_IS_STARTED_IN_ENGINE) &&
+		     txn_has_flag(txn, TXN_SUPPORTS_MVCC) != has_mvcc)) {
+		struct engine *other = NULL;
+		for (int i = 0; i < MAX_TX_ENGINE_COUNT; i++) {
+			if (txn->engines[i] != NULL &&
+			    !(txn->engines[i]->flags &
+			      ENGINE_SUPPORTS_MVCC) == has_mvcc) {
```

Comment:
Nit: I believe that `(txn->engines[i]->flags & ENGINE_SUPPORTS_MVCC) != has_mvcc` would be a bit easier to understand.

Reply:
It wouldn't work because `foo & bar` returns the result of bit-wise "and" while a boolean flag stores 0 or 1 so negation is necessary for correctness in this case.